cas-server-support-saml-idp/src/main/java/org/jasig/cas/support/saml/web/idp/profile/AbstractSamlProfileHandlerController.java

```diff
@@ -0,0 +1,368 @@
+     * @return the registered service and verify
+     */
+    protected SamlRegisteredService verifySamlRegisteredService(final AuthnRequest authnRequest) {
+        final String serviceId = authnRequest.getAssertionConsumerServiceURL();
+        logger.debug("Checking service access in CAS service registry for [{}]", serviceId);
+        final RegisteredService registeredService =
+                this.servicesManager.findServiceBy(this.webApplicationServiceFactory.createService(serviceId));
+        if (registeredService == null || !registeredService.getAccessStrategy().isServiceAccessAllowed()) {
+            logger.warn("[{}] is not found in the registry or service access is denied. Ensure service is registered in the CAS registry",
+                    serviceId);
+            throw new UnauthorizedServiceException(UnauthorizedServiceException.CODE_UNAUTHZ_SERVICE);
+        }
+
+        if (registeredService instanceof SamlRegisteredService) {
+            final SamlRegisteredService samlRegisteredService = (SamlRegisteredService) registeredService;
+            logger.debug("Located SAML service in the registry as [{}] with the metadata location of [{}]",
+                    samlRegisteredService.getServiceId(), samlRegisteredService.getMetadataLocation());
+            return samlRegisteredService;
+        }
+        logger.error("Service [{}] is found in registry but it is not defined as a SAML service", serviceId);
+        throw new UnauthorizedServiceException(UnauthorizedServiceException.CODE_UNAUTHZ_SERVICE);
+    }
+
+    /**
+     * Initialize callback service.
+     *
+     * @param callbackUrl the callback url
+     * @return the service
+     */
+    protected Service registerCallback(final String callbackUrl) {
+        final Service callbackService = webApplicationServiceFactory.createService(this.casServerPrefix.concat(callbackUrl));
+        logger.debug("Initialized callback service [{}]", callbackService);
+
+        if (!servicesManager.matchesExistingService(callbackService)) {
+            final RegexRegisteredService service = new RegexRegisteredService();
+            service.setId(new SecureRandom().nextLong());
+            service.setName(service.getClass().getSimpleName());
+            service.setDescription(SamlProtocolConstants.ENDPOINT_SAML2_SSO_PROFILE_POST_CALLBACK.concat(" Callback Url"));
+            service.setServiceId(callbackService.getId());
+
+            logger.debug("Saving callback service [{}] into the registry", service);
+            this.servicesManager.save(service);
+            this.servicesManager.reload();
+        }
+        return callbackService;
+    }
+
+    /**
+     * Store authn request.
+     *
+     * @param request      the request
+     * @param authnRequest the authn request
+     */
+    protected void storeAuthnRequest(final HttpServletRequest request, final AuthnRequest authnRequest) {
+        final HttpSession session = request.getSession();
```

Comment:
The OAuth protocol is not related to web sessions. But our current implementation is because it was easier developing that way. Maybe I'll revisit that as well.